### PR TITLE
test(openapi-v3-types): improve test coverage

### PR DESCRIPTION
**Pull request by virkt25**
connected to #181 

set up the test infrastructure for this package as well as write tests for coverage.

NOTE: Interfaces are meant to be tested by just having the files compile as the typing information is otherwise lost afterwards.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by raymondfeng** [comment]
The CI fails due to the following error, which is pretty weird.
```
> @loopback/testlab@0.5.3 build /home/travis/build/strongloop/loopback-next/packages/testlab
> lb-tsc es2017
packages/testlab/src/validate-api-spec.ts(8,9): error TS2305: Module '"util"' has no exported member 'promisify'.
```

**Comment by virkt25** [comment]
Weird, we actually ran into this issue once last week and spent an hour trying to debug to no avail and the issue just cleared it self up again

**Comment by raymondfeng** [comment]
I was able to reproduce the problem after a fresh checkout and npm cache clean --force.

Here is the issue:
```
@loopback/testlab@0.5.3 /private/tmp/loopback-next/packages/testlab
├─┬ @loopback/build@0.4.2 -> /private/tmp/loopback-next/packages/build
│ ├── @types/node@8.10.4 
│ └─┬ strong-docs@1.10.2
│   └─┬ typedoc@0.11.1
│     ├─┬ @types/fs-extra@5.0.1
│     │ └── @types/node@8.10.4  deduped
│     └─┬ @types/shelljs@0.7.8
│       ├─┬ @types/glob@5.0.35
│       │ └── @types/node@8.10.4  deduped
│       └── @types/node@8.10.4  deduped
├─┬ @types/fs-extra@5.0.1
│ └── @types/node@7.0.60 
├─┬ @types/shot@3.4.0
│ └── @types/node@7.0.60  deduped
└─┬ @types/supertest@2.0.4
  └─┬ @types/superagent@3.5.7
    └── @types/node@7.0.60  deduped
```

The `testlab` package ends up with `@types/node@7.x`.

**Comment by virkt25** [comment]
Thanks! That actually explains a lot. `@types/superagent` sets the `@types/node` to `*` which results in the latest to be installed ... and `@types/node` did a release for `7.0.60` 3 hours ago hence everything breaking.

I tried setting `@types/node` as a devDependecy for `@loopback/testlab` and that package builds but this has to be done for every package using `@loopback/testlab` individually (for some reason). 

Any ideas for a simpler way around this?

**Comment by raymondfeng** [comment]
Hmm, `*` is resolved as the latest version published (not the greatest version).

**Comment by raymondfeng** [comment]
I think we cannot rely on the `loopback-next` root level `@types/node` version any more. It's probably safe to add `"@types/node": "^8.10.4"` to all packages that require 8.0 apis.

File: packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  ReferenceObject,
+  DiscriminatorObject,
+  XMLObject,
+  ExternalDocumentationObject,
+  ISpecificationExtension,
+  createEmptyApiSpec,
+  OpenAPIObject,
+} from '../..';
+
+describe('openapi-v3-types unit tests', () => {
```
Comment (shimks) [inline packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts]: Looking at the tests written below, wouldn't it be more accurate to describe this `describe` block as `createEmptyApiSpec unit tests`?

File: packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    const serverUrl = server ? server.url : '';
+    expect(serverUrl).to.equal('/');
+  });
+
+  function createEmptySpec() {
+    emptySpec = createEmptyApiSpec();
+  }
+});
+
+/**
```
Comment (shimks) [inline packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts]: I think it'd make things clearer that the classes below are tests if they were enclosed within a describe block. I'm not sure if a describe block is allowed to exist without any `it` blocks though

File: packages/openapi-v3-types/test/unit/type-guards.unit.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      expect(isReferenceObject(referenceObject)).to.be.True();
+    });
+
+    it('returns false for a schema object', () => {
+      const schemaObject = new TestSchemaObject();
+      expect(isReferenceObject(schemaObject)).to.be.False();
+    });
+  });
+
+  class TestSchemaObject implements SchemaObject {}
```
Comment (shimks) [inline packages/openapi-v3-types/test/unit/type-guards.unit.ts]: I don't know too much about `SchemaObject`, but shouldn't the test here be more than just an empty class?

File: packages/testlab/src/validate-api-spec.ts [hunk 4]
```diff
@@ -3,13 +3,13 @@
...
 
-import {OpenApiSpec} from '@loopback/openapi-v3-types';
 const validator = require('swagger2openapi/validate.js');
 import {promisify} from 'util';
 
 const validateAsync = promisify(validator.validate);
 
-export async function validateApiSpec(spec: OpenApiSpec): Promise<void> {
+// tslint:disable-next-line:no-any
+export async function validateApiSpec(spec: any): Promise<void> {
```
Comment (jannyHou) [inline packages/testlab/src/validate-api-spec.ts]: hmm, any reason for changing the type of `spec`?
Assume the validator works fine, either `OpenApiSpec` or `any` works, while `OpenApiSpec` can catch error when editing I think.

File: packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts (hunk 2 above)
Comment (virkt25) [reply packages/openapi-v3-types/test/unit/openapi-v3-spec-types.unit.ts]: I can see if that works

File: packages/openapi-v3-types/test/unit/type-guards.unit.ts (hunk 3 above)
Comment (virkt25) [reply packages/openapi-v3-types/test/unit/type-guards.unit.ts]: Nope, `SchemaObject` doesn't have any required properties

**Comment by virkt25** [comment]
@shimks There are a few other interfaces we define but they reply on `MapObject` and implement a generic extensible interface. Not sure if there's any value in that since a generic interface doesn't have any properties it requires be present on a class ... which is what the other Test Classes are validating.

**Comment by shimks** [comment]
I don't think so considering the whole v3 to v2 conversion story was dropped from dp2.

**Pull request merged**